### PR TITLE
HDDS-8418. Lease Manager unsafe method fix

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

registerCallback in Lease is not safe if done after lease expiry in parallel, so this callback will never be registered. Removed this method exposer.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8418

## How was this patch tested?

UT case and IT for checking any impact

### CODE REVIEW

**Comment by adoroszlai** [comment]
@sumitagrawl [there are](https://github.com/sumitagrawl/ozone/actions/runs/4720770840) multiple test failures, please check

**Review by adoroszlai — COMMENTED** [review]
@sumitagrawl thanks for working on this.  It's not clear what the problem was and how the patch is supposed to fix it.

I suspect [intermittent failure in `TestCloseContainerEventHandler`](https://github.com/apache/ozone/actions/runs/4837657037/jobs/8621613046#step:6:1508) could be the result of the problem being fixed:

```
[INFO] Running org.apache.hadoop.hdds.scm.container.TestCloseContainerEventHandler
Error:  Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.357 s <<< FAILURE! - in org.apache.hadoop.hdds.scm.container.TestCloseContainerEventHandler
Error:  org.apache.hadoop.hdds.scm.container.TestCloseContainerEventHandler.testCloseContainerWithDelayByLeaseManager  Time elapsed: 4.528 s  <<< FAILURE!
Wanted but not invoked:
eventPublisher.fireEvent(
    TypedEvent{payloadType=CommandForDatanode, name='Datanode_Command'},
    <Capturing argument>
);
-> at org.apache.hadoop.hdds.scm.container.TestCloseContainerEventHandler.testCloseContainerWithDelayByLeaseManager(TestCloseContainerEventHandler.java:151)
Actually, there were zero interactions with this mock.
```

Synchronization in `Lease` is very mixed after this patch:

* write to `expired` is synchronized (`invalidate()`), but read is not (`hasExpired()`), except when called from other synchronized method
* `callbacks` is a `synchronizedList`, item addition is also synchronized on `Lease` (`registerCallback()`), but `getCallbacks()` returns the list as is, and iteration happens without either lock
* `leaseTimeout` is atomic, write is also synchronized (`renew()`), but read is not (`getLeaseLifeTime()`)

Several methods called from `LeaseManager` are also protected by synchronization in that class.

Instead of adding a few `synchronized` keywords here and there, it would be great to make it consistent.

**Comment by szetszwo** [comment]
The bug is the `acquire` methods returning a `Lease` and then `registerCallBack` later.  It seems the current change does not prevent that.  We should pass the callbacks to `acquire` at the same time in order to fix the bug.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/events/EventWatcher.java [hunk 1]
```diff
@@ -157,13 +146,14 @@ protected synchronized void handleCompletion(COMPLETION_PAYLOAD
     }
   }
 
-  private synchronized void handleTimeout(EventPublisher publisher,
+  private synchronized Void handleTimeout(EventPublisher publisher,
```
Comment (sadanand48) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/events/EventWatcher.java:149]: I didn't get this part, what is the impact of changing void-> Void and returning null

Comment (sumitagrawl) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/events/EventWatcher.java:149]: Callback as getting registered has return type is Callback with "Void" Object, so this gives error if return type mismatch,

Syntax for acquire: `public synchronized Lease<T> acquire(
      T resource, Callable<Void> callback)`

Below is error if we keep void:
`Bad return type in lambda expression: void cannot be converted to Void`

Comment (sadanand48) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/events/EventWatcher.java:149]: Got it , looks like autoboxing won't happen for void type. Thanks for explaining

**Comment by sadanand48** [comment]
Thanks @sumitagrawl for the fix, @adoroszlai @szetszwo for reviewing.

**Pull request merged**